pkgs/development/compilers/llvm/common/default.nix

```diff
@@ -1058,7 +1059,13 @@ let
         else
           libraries.compiler-rt-libc;
 
-      stdenv = overrideCC stdenv buildLlvmTools.clang;
+      stdenv =
+        let
+          stdenv' = overrideCC stdenv buildLlvmTools.clang;
+        in
+        # Treat libc++ as a “system” library on Darwin where all versions use the same underlying libc++.
+        # Use libcxxStdenv to use a specific version of LLVM and libc++ regardless.
+        if stdenv.targetPlatform.isDarwin then overrideLibcxx stdenv' else stdenv';
```

Comment:
so i believe that the ABI is backwards compatible but there are new symbols that get added to the new ABI which are not available in the older one. eg -- trying to compile against libc++ from llvm 18 and then only providing llvm 16 libc++ will break.

Reply:
Latest push makes this conditional based on version.

```shell
$ nix-instantiate . -A libcxx
/nix/store/z4lm72bh3ysll6dxr3zfrdgq34hdryb4-libcxx-16.0.6.drv
$ nix-instantiate . -A llvmPackages_16.stdenv.cc.libcxx
/nix/store/z4lm72bh3ysll6dxr3zfrdgq34hdryb4-libcxx-16.0.6.drv
$ nix-instantiate . -A llvmPackages_19.stdenv.cc.libcxx
/nix/store/diq1vrfb58lpmywf4lafip71wlxjhk04-libcxx-19.1.1.drv
$ nix-instantiate . -A llvmPackages_13.stdenv.cc.libcxx
/nix/store/bzlk94jnvrf4ckii1flljy8pij0kvbin-libcxx-16.0.6-13.0.1.drv
```